librz/il/rzil_vm.c

```diff
@@ -0,0 +1,460 @@
+	rz_return_val_if_fail(name, NULL);
+	HtPP *lbl_table = vm->vm_global_label_table;
+
+	RzILEffectLabel *lbl = rz_il_effect_new_label(name, EFFECT_LABEL_ADDR);
+	lbl->addr = NULL;
+	ht_pp_insert(lbl_table, name, lbl);
+
+	return lbl;
+}
+
+/**
+ * Update the address info of a label
+ * \param vm RzILVM, pointer to VM
+ * \param name string, name of this label
+ * \return lbl RzILEffectLabel, pointer to label instance
+ */
+RZ_API RzILEffectLabel *rz_il_vm_update_label(RzILVM *vm, RZ_NONNULL char *name, RzILBitVector *addr) {
+	rz_return_val_if_fail(name, NULL);
+	RzILEffectLabel *lbl = ht_pp_find(vm->vm_global_label_table, name, NULL);
+	if (lbl->addr) {
+		rz_il_bv_free(lbl->addr);
+	}
+	lbl->addr = rz_il_bv_dup(addr);
+
+	return lbl;
+}
+
+/**
+ * Store an opcode list to address
+ * \param vm RzILVM, pointer to VM
+ * \param addr RzILBitVector, address of this opcode list
+ * \param oplist RzPVector of RzILOp, core theory opcodes
+ */
+RZ_API void rz_il_vm_store_opcodes_to_addr(RzILVM *vm, RzILBitVector *addr, RzPVector *oplist) {
+	rz_return_if_fail(addr && oplist);
+	ht_pp_insert(vm->ct_opcodes, addr, oplist);
+}
+
+/**
+ * Make a core theory opcode vector
+ * \param num int, number of total opcodes
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist(int num, ...) {
+	va_list args;
+	RzILOp *cur_op;
+	RzPVector *oplist = rz_pvector_new((RzPVectorFree)rz_il_free_op);
+
+	va_start(args, num);
+	for (int i = 0; i < num; ++i) {
+		cur_op = va_arg(args, RzILOp *);
+		rz_pvector_push(oplist, cur_op);
+	}
+	va_end(args);
+
+	return oplist;
+}
+
+// WARN : convertion breaks the original data
+static RzILBool *bitv_to_bool(RzILBitVector *bitv) {
+	RzILBool *result;
+	result = rz_il_new_bool(!rz_il_bv_is_zero_vector(bitv));
+	rz_il_bv_free(bitv);
+	return result;
+}
+
+static RzILVal *bitv_to_val(RzILBitVector *bitv) {
+	RzILVal *ret = rz_il_new_value();
+	ret->type = RZIL_VAR_TYPE_BV;
+	ret->data.bv = bitv;
+	return ret;
+}
+
+static RzILVal *bool_to_val(RzILBool *b) {
+	RzILVal *ret = rz_il_new_value();
+	ret->type = RZIL_VAR_TYPE_BOOL;
+	ret->data.b = b;
+	return ret;
+}
+
+static RzILBitVector *val_to_bitv(RzILVal *val) {
+	RzILBitVector *ret;
+	if (val->type != RZIL_VAR_TYPE_BV) {
+		RZ_LOG_ERROR("RzIL : Expected bool, but UNK/BOOL detected\n");
+		return NULL;
+	}
+
+	ret = val->data.bv;
+	RZ_FREE(val);
+	return ret;
+}
+
+static RzILBool *val_to_bool(RzILVal *val) {
+	RzILBool *ret;
+	if (val->type != RZIL_VAR_TYPE_BOOL) {
+		if (val->type == RZIL_VAR_TYPE_BV) {
+			return bitv_to_bool(val_to_bitv(val));
+		}
+		RZ_LOG_ERROR("RzIL : Expected bool, but UNK detected\n");
+		return NULL;
+	}
+
+	ret = val->data.b;
+	RZ_FREE(val);
+	return ret;
+}
+
+// VM auto convertion
+RzILBitVector *rz_il_evaluate_bitv(RzILVM *vm, RzILOp *op, RZIL_OP_ARG_TYPE *type) {
```

Comment:
can this be explained a bit better in the doc? What does it actually do? Sorry for the maybe stupid question.

Reply:
It evaluate an expression and ensure you to get a bitvector. When evaluating a sub expression, we may get a bool/bv/val value, but the opcode may require a bitvector operand. This function will convert valid types to a bitvector.